### ISSUE TITLE
Logstash 6.1.0: Beats input chokes on empty string JSON keys in events shipped with filebeat 

### ISSUE THREAD

**Issue by lukasgraf** [elastic/logstash#8852]
After upgrading logstash from 6.0.0 to 6.1.0, transport of events with **empty string JSON keys** from **filebeat** LS output to the **logstash beats input** doesn't work anymore (`Exception: -1` or `Exception: null` when filebeat client tries to establish TCP connection):

```
[INFO ][org.logstash.beats.BeatsHandler] [local: 172.26.0.2:5044, remote: 172.26.0.5:53834] Exception: -1
```
(See full logs below).

The problem seems to be caused by data that has the **empty string** as a JSON key, like
```
{"data": {"foo": 1, "": 2}}
```
when using [filebeat JSON decoding](https://www.elastic.co/guide/en/beats/filebeat/current/configuration-filebeat-options.html#config-json).

---

I first noticed the issue after upgrading the entire stack from 6.0 to 6.1. In an attempt to isolate the problem I updated *just* logstash to 6.1.0 (kept rest of the stack at 6.0.0). The problem persists in that configuration. If I keep logstash at 6.0.0 and update filebeat to 6.1.0, everything works. So I assume the problem is related to a change in LS 6.1.0.

The messages from the log below keep repeating indefinitely. It seems the filebeat client fails to establish a connection to the beats input.

### Sample event
```
{"data": {"foo": 1, "": 2}}
```

### Steps to reproduce

I've created a repo including a docker-compose file, sample data and configurations that trivially demonstrates the problem (and the fact that this used to work on 6.0.0): https://github.com/lukasgraf/logstash-8852-testcase

### Filebeat config
```yaml
output.logstash:
  hosts: ["logstash-ingest:5044"]
logging.level: debug
```
(`logstash-ingest` is the docker-managed DNS hostname of the host where logstash beats input is running). 

### Logstash pipeline input config
```ruby
input {
    beats {
        port => "5044"
    }
}
```

### Versions
- Filebeat: `6.1.0`
- Logstash: `6.1.0` (before issue started appearing: `6.0.0`)
- logstash-input-beats: `5.0.3` (before issue started appearing: `5.0.2`) *(as shipped in the official docker images for the respective LS versions)*

### Environment
- OS: Mac OS 10.11.6
- Docker for Mac, using official Elastic Docker images from `docker.elastic.co`.
- Topology: Filebeat and Logstash on different docker hosts, logstash beats input listening on public interface `0.0.0.0:5044`

### Logs

These are the logs with log level DEBUG for both filebeat and logstash.inputs.beats. This is a sample of a handful connection attempts, the same block of messages keeps repeating indefinitely. The specific exception seems to vary: If seen both `Exception: -1` and `Exception: null`. 

```
filebeat_1           | 2017/12/17 16:29:15.444704 output.go:92: ERR Failed to publish events: client is not connected
filebeat_1           | 2017/12/17 16:29:15.444779 async.go:94: DBG [logstash] connect
filebeat_1           | 2017/12/17 16:29:15.482378 async.go:142: DBG [logstash] 2048 events out of 2048 events sent to logstash host logstash-ingest:5044. Continue sending
logstash-ingest_1    | [2017-12-17T16:29:15,515][INFO ][org.logstash.beats.BeatsHandler] [local: 172.26.0.2:5044, remote: 172.26.0.5:53832] Exception: -1
filebeat_1           | 2017/12/17 16:29:15.516385 client.go:201: DBG [transport] handle error: read tcp 172.26.0.5:53832->172.26.0.2:5044: read: connection reset by peer
filebeat_1           | 2017/12/17 16:29:15.516536 async.go:235: ERR Failed to publish events caused by: read tcp 172.26.0.5:53832->172.26.0.2:5044: read: connection reset by peer
filebeat_1           | 2017/12/17 16:29:15.516621 client.go:114: DBG [transport] closing
filebeat_1           | 2017/12/17 16:29:15.545998 async.go:142: DBG [logstash] 2048 events out of 2048 events sent to logstash host logstash-ingest:5044. Continue sending
filebeat_1           | 2017/12/17 16:29:15.546087 async.go:99: DBG [logstash] close connection
filebeat_1           | 2017/12/17 16:29:15.546147 async.go:99: DBG [logstash] close connection
filebeat_1           | 2017/12/17 16:29:15.546203 async.go:235: ERR Failed to publish events caused by: client is not connected

filebeat_1           | 2017/12/17 16:29:16.546573 output.go:92: ERR Failed to publish events: client is not connected
filebeat_1           | 2017/12/17 16:29:16.546656 async.go:94: DBG [logstash] connect
filebeat_1           | 2017/12/17 16:29:16.576110 async.go:142: DBG [logstash] 2048 events out of 2048 events sent to logstash host logstash-ingest:5044. Continue sending
logstash-ingest_1    | [2017-12-17T16:29:16,593][INFO ][org.logstash.beats.BeatsHandler] [local: 172.26.0.2:5044, remote: 172.26.0.5:53834] Exception: -1
filebeat_1           | 2017/12/17 16:29:16.594678 client.go:201: DBG [transport] handle error: read tcp 172.26.0.5:53834->172.26.0.2:5044: read: connection reset by peer
filebeat_1           | 2017/12/17 16:29:16.594753 async.go:235: ERR Failed to publish events caused by: read tcp 172.26.0.5:53834->172.26.0.2:5044: read: connection reset by peer
filebeat_1           | 2017/12/17 16:29:16.594788 client.go:114: DBG [transport] closing
filebeat_1           | 2017/12/17 16:29:16.627035 async.go:142: DBG [logstash] 2048 events out of 2048 events sent to logstash host logstash-ingest:5044. Continue sending
filebeat_1           | 2017/12/17 16:29:16.627100 async.go:99: DBG [logstash] close connection
filebeat_1           | 2017/12/17 16:29:16.627144 async.go:99: DBG [logstash] close connection
filebeat_1           | 2017/12/17 16:29:16.627295 async.go:235: ERR Failed to publish events caused by: client is not connected
filebeat_1           | 2017/12/17 16:29:17.628374 output.go:92: ERR Failed to publish events: client is not connected
filebeat_1           | 2017/12/17 16:29:17.628483 async.go:94: DBG [logstash] connect
filebeat_1           | 2017/12/17 16:29:17.669998 async.go:142: DBG [logstash] 2048 events out of 2048 events sent to logstash host logstash-ingest:5044. Continue sending
logstash-ingest_1    | [2017-12-17T16:29:17,683][INFO ][org.logstash.beats.BeatsHandler] [local: 172.26.0.2:5044, remote: 172.26.0.5:53836] Exception: -1
filebeat_1           | 2017/12/17 16:29:17.685046 client.go:201: DBG [transport] handle error: read tcp 172.26.0.5:53836->172.26.0.2:5044: read: connection reset by peer
filebeat_1           | 2017/12/17 16:29:17.685449 async.go:235: ERR Failed to publish events caused by: read tcp 172.26.0.5:53836->172.26.0.2:5044: read: connection reset by peer
filebeat_1           | 2017/12/17 16:29:17.685586 client.go:114: DBG [transport] closing
filebeat_1           | 2017/12/17 16:29:17.721855 async.go:142: DBG [logstash] 2048 events out of 2048 events sent to logstash host logstash-ingest:5044. Continue sending
filebeat_1           | 2017/12/17 16:29:17.721911 async.go:99: DBG [logstash] close connection
filebeat_1           | 2017/12/17 16:29:17.721929 async.go:99: DBG [logstash] close connection
filebeat_1           | 2017/12/17 16:29:17.722014 async.go:235: ERR Failed to publish events caused by: client is not connected
filebeat_1           | 2017/12/17 16:29:18.722897 output.go:92: ERR Failed to publish events: client is not connected
filebeat_1           | 2017/12/17 16:29:18.722993 async.go:94: DBG [logstash] connect
filebeat_1           | 2017/12/17 16:29:18.764418 async.go:142: DBG [logstash] 2048 events out of 2048 events sent to logstash host logstash-ingest:5044. Continue sending
logstash-ingest_1    | [2017-12-17T16:29:18,780][INFO ][org.logstash.beats.BeatsHandler] [local: 172.26.0.2:5044, remote: 172.26.0.5:53838] Exception: -1
filebeat_1           | 2017/12/17 16:29:18.780962 client.go:201: DBG [transport] handle error: read tcp 172.26.0.5:53838->172.26.0.2:5044: read: connection reset by peer
filebeat_1           | 2017/12/17 16:29:18.781080 async.go:235: ERR Failed to publish events caused by: read tcp 172.26.0.5:53838->172.26.0.2:5044: read: connection reset by peer
filebeat_1           | 2017/12/17 16:29:18.781107 client.go:114: DBG [transport] closing
filebeat_1           | 2017/12/17 16:29:18.812404 async.go:142: DBG [logstash] 2048 events out of 2048 events sent to logstash host logstash-ingest:5044. Continue sending
filebeat_1           | 2017/12/17 16:29:18.812451 async.go:99: DBG [logstash] close connection
filebeat_1           | 2017/12/17 16:29:18.812474 async.go:235: ERR Failed to publish events caused by: client is not connected
filebeat_1           | 2017/12/17 16:29:18.812709 async.go:99: DBG [logstash] close connection
...
```

**Comment by jsvd** [comment]
This seems to have been already detected and fixed in https://github.com/elastic/logstash/pull/8824 for the future 6.2 and 7.0 releases, I've asked Jake to backport this to 6.1.x

**Comment by lukasgraf** [comment]
Excellent, thank you!

I'm also well aware that it makes no sense to attempt to store JSON with empty string keys (even though that's technically valid JSON). But real world data is *messy* 😉 .

So I previously hit this problem where Elasticsearch rejected documents like these - which is why I dropped the empty keys using a filter in logstash. But now that logstash already rejects them I'm in quite a pickle, so the backport would be very much welcomed. 💚

**Comment by jakelandis** [comment]
@lukasgraf - Thanks for reporting this, and super thanks ❤️ for the details and reproduction case. 

I tested before and after applying the fix for #8824 and was able to reproduce and then fix the issue with the patch. 

I am going to close this as fixed with `6.1.1`.

**Comment by lukasgraf** [comment]
@jakelandis awesome, thanks a lot for the quick backport!
